proposals/nnnn-trailing-comma-lists.md

```diff
@@ -0,0 +1,294 @@
+}
+
+arrayOfS.sorted(
+  \.a, 
+  \.b,
+//  \.c
+) // ❌ Unexpected ',' separator
+```
+
+Since [#21381](https://github.com/apple/swift/pull/21381) has been merged back in 2019 **enum associated values** supports default values and are a good fit for trailing comma as well.
+
+**Multiple conditions** in `if`, `guard` and `while` are also comma-separated list and add, remove, reorder and comment in/out are not uncommon practice during development.
+
+```swift
+if 
+   condition1,
+   condition2,
+//   condition3
+{ // ❌ Cannot convert value of type '() -> ()' to expected condition type 'Bool'
+                    
+} // ❌ Expected '{' after 'if' condition
+```
+
+These benefits can be extended to other comma-separated lists found in the Swift language.
+
+### Code Generation
+
+**Plugins** and **Macros** have made it possible to generate code using swift and trailing comma would allow generate lists without worrying about a special condition for the last element.
+
+### Code Diff
+
+A tangential motivation is that trailing comma makes version-control diffs cleaner.
+
+Without trailing comma:
+```diff
+foo(
+-  a: Int
++  a: Int,
++  b: Int
+)
+```
+With trailing comma:
+```diff
+foo(
+  a: Int,
++  b: Int,
+)
+```
+
+> [!NOTE]  
+> A similar proposal was [rejected](https://forums.swift.org/t/rejected-se-0084-allow-trailing-commas-in-parameter-lists-and-tuples/2777) back in 2016 for Swift 3. It's been 8 years since that, the swift language has evolved a lot, some changes highlighted above as motivation, and the code style that "puts the terminating right parenthesis on a line following the arguments to that call" has been widely adopted by community, swift standard library codebase, swift-format, docc documentation and Xcode. Therefore, not encourage or endorse this code style doesn't hold true anymore nor is a reason for rejection.
+
+## Proposed solution
+
+This proposal adds support for trailing comma for comma-separated lists, but only when there's a clear terminator. A enum case label list won't support, for example:
+
+```swift
+enum E {
+  case a, b, c, // ❌ Expected identifier after comma in enum 'case' declaration
+}
+```
+
+Trailing comma will be allowed in the following comma-separated lists:
+
+- Tuples and tuple patterns 
+
+```swift
+(1, 2,)
+let block: (Int, Int,) -> Void = { (a, b,) in  }
+let (a, b,) = (1, 2,)
+for (a, b,) in zip(s1, s2) { }
+```
+
+**Trailing comma will be allowed in single-element lists but not in zero-element lists.**
+
+```swift
+(1,) // OK
+(,) // ❌ expected value in tuple
+```
+
+- Initializers, functions and associated values arguments and parameters
+
+```swift
+
+m[x, y,]
+
+func foo(a: Int, b: Int,) { }
+
+struct S {
+    init(a: Int, b: Int,) { }
+}
+
+enum E {
+    case foo(a: Int, b: Int,)
+}
+```
+- Subscripts
+
+```swift
+let value = m[x, y,]
+```
+
+- `KeyPath` subscripts 
+
+```swift
+let keyPath = \Foo.bar[x,y,]
+          
+f(\.[x,y,])
+```
+
+- Attributes arguments
```

Comment:
Similar comment for this and the following. I understand that you've had to handle separate cases in the implementation, but to end users they're all argument lists, I think.

It would seem logical that use sites for expression macros including built-in ones such as `#colorLiteral(...)`, attached macros, property wrappers, and built-in attributes all would fit under one heading of "argument lists"—although of course it is important to spell it out.